System/Metrics/Dimensional.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+
+-- | This module defines a type for mutable dimensional string-valued label.
+-- Dimensional label are variable values and can be used to track e.g. the
+-- current number of call to an api endpoint by its http return code.
+-- All operations on Dimensional label are thread-safe.
+module System.Metrics.Dimensional where
+
+import qualified System.Metrics.Counter as Counter
+import qualified System.Metrics.Distribution as Distribution
+import qualified System.Metrics.Gauge as Gauge
+import qualified System.Metrics.Label as Label
+
+import Control.Applicative ((<$>), pure)
+import Control.Exception (throwIO)
+import Data.HashMap.Strict (HashMap, empty)
+import qualified Data.HashMap.Strict as HashMap
+import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
+import Data.Text (Text)
+import Data.Typeable
+import GHC.Exception (Exception)
+import Prelude hiding (lookup)
+
+type Name = Text
+
+type Explanation = Text
```

Comment:
This is a bit more of an overall comment but I think all metrics should support an explanation (although I haven't given thought to how to retrofit this into the current API) and each dimension could also use an explanation.

In general I think one might want to attach the following metadata to a metric:
* It's type e.g. "seconds" or "megabytes". This helps when automatically creating graphs and UIs for metrics.
* A human-readable explanation of the metrics.
* An explanation of each dimension.